Fix logging and metrics for TrackingKafkaConsumer with partition assignment

AndreKurait opened the pull request:
### Description
Improves logging in TrackingKafkaConsumer by improving clarity when an empty partition collection is assigned/revoked and correcting the behavior with `this` is logged by adding a toString method. Fixes metrics by only emitting metrics when a nonEmpty partition collection is modified and corrected for negative delta when partitions are revoked. Also moved a couple Warn Messages to Info for partition assginment/revocation.
* Category: Bug Fix
* Why these changes are required? Improving correctness of observability in the replayer
* What is the old behavior before changes and new behavior after changes? Old Behavior, partition assignment increased during revocations and some logs were unintelligible, e.g.: 
```
[WARN ] 2024-03-08 17:42:18,263 [kafkaConsumerThread-1-1] TrackingKafkaConsumer - org.opensearch.migrations.replay.kafka.TrackingKafkaConsumer@597f4cadpartitions added for
```
Now metrics are corrected and log messages are meaningful. 


### Issues Resolved
N/A

Is this a backport? If so, please add backport PR # and/or commits #
No

### Testing
Unit Testing and PR Auto Tests

### Check List
- [x ] New functionality includes testing
  - [ x] All tests pass, including unit test, integration test and doctest
- [ x] New functionality has been documented
- [ x] Commits are signed per the DCO using --signoff

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.
For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/opensearch-project/OpenSearch/blob/main/CONTRIBUTING.md#developer-certificate-of-origin).

gregschohn commented on TrafficCapture/trafficReplayer/src/main/java/org/opensearch/migrations/replay/tracing/KafkaConsumerContexts.java:
```diff
@@ -59,7 +59,7 @@ private MetricInstruments(Meter meter) {
         public void onPartitionsRevoked(Collection<TopicPartition> partitions) {
             meterIncrementEvent(getMetrics().kafkaPartitionsRevokedCounter);
-            onParitionsAssignedChanged(partitions.size());
+            onParitionsAssignedChanged(-1 * partitions.size());
```
Could you fix the spelling of this function? M

AndreKurait replied:
Addressed

gregschohn reviewed (commented):
For now, when a partition is revoked, it may signal an error or serious performance issue elsewhere. Once we support parallel replacers, I agree that the log level for partition changes should go to info, but until then, I'm more comfortable with it being at warn. I could be convinced otherwise though. Other than that and one spelling error this LGTM

AndreKurait commented:
> For now, when a partition is revoked, it may signal an error or serious performance issue elsewhere. Once we support parallel replacers, I agree that the log level for partition changes should go to info, but until then, I'm more comfortable with it being at warn. I could be convinced otherwise though. Other than that and one spelling error this LGTM

Addressed and moved to warn
